src/nsmd.cpp

```diff
@@ -365,13 +380,21 @@ handle_client_process_death ( int pid )
 
 void handle_sigchld ( )
 {
+    // compare waitpid(2)
     for ( ;; )
     {
-        int status;
-        pid_t pid = waitpid(-1, &status, WNOHANG);
+        int status = 1; // make it not NULL to enable information storage in status
+        pid_t pid = waitpid(-1, &status, WNOHANG); //-1 meaning wait for any child process. pid_t is signed integer
 
         if (pid <= 0)
-            break;
+            break; // no child process has ended this loop. Check again.
+
+        //One child process has stopped. Find which and figure out the stop-conditions
```

Comment:
This part of the code is a bit off..
1. not checking if the returned client pointer is valid (it could be null)
2. code checks for everything regardless of current state. please either add a check for point 1 and then check the exit status inside that scope, or check the exit status first and inside that scope try to get the client by pid
3. -1 is a common error code, I believe you will get the same behaviour if an application crashes (but I am not sure of this, we need to check). please try with an application that purposefully crashes after some time running. My suspicion is that crashed clients will trigger the same conditions as the one you are checking for here.

Reply:
This is original nsmd code.  We can test this, but I would prefer not doing that in this PR.

Reply:
In case you meant the code below: 

1./2.  Will test for NULL first.
3. Tested with Patroneo through provoked python error and ctypes segfault. Also with kluppe (not an nsm client) which segfaults as well.  This will, as intended, end in just the stopped status.